### DISCUSSION TITLE
ESLint / Prettier を Biome に置き換える

### DISCUSSION THREAD

**Discussion by m1sk9** [pulsate-dev/pulsate#568] [Ideas]
現在 Pulsate, Pulsate-Web では ESLint, Prettier を使用しているが, これを Biome に置き換えたい.
Biome は CLI での Migrate に対応している.

[Migrate from ESLint and Prettier | Biome](https://biomejs.dev/guides/migrate-eslint-prettier/)

### pros

- Rust を採用しており, ESLint や Prettier よりも高速 (検証はできていない. 理由は後述)
- `.jsonc` による設定ファイルをサポートしており, コメントが利用できる ( `.json` にも対応)
- Renovate にルールセットが用意されており, `biome.jsonc` 内の `$schema` を最新に保つ事ができる.
- `biome ci` という GitHub Action などの CI 環境での実行が想定されているサブコマンドが用意されている. 

### cons

- YAML や Markdown など一部言語ではサポートが行われていなかったり, 部分サポートになっている: [Language support | Biome](https://biomejs.dev/internals/language-support/ )
- ESLint の移行ができない (Pulsate において)
	- Pulsate では ESLint v9 でサポートされた Flat Config を使用しているが, おそらくこれが原因? ただしドキュメントでは新しい形式でのコンフィグもサポートしているようなので不具合な可能性が否定できないが GitHub にそれらしい Issue がない.

**Comment by Allianaab2m** [comment]
> ESLint の移行ができない (Pulsate において)

ESLintの外部Pluginで実現していた部分がBiome本体に含まれていることが確認できれば，migrateせずともinitして導入しても構わないと思います．

**Comment by m1sk9** [comment]
> biome ci という GitHub Action などの CI 環境での実行が想定されているサブコマンドが用意されている.

この時点で相当 Biome に軍配があがる. というかルールとかを `pnpm add` などで別途入れる必要がないのも大きい.

https://github.com/pulsate-dev/pulsate/blob/3b8fbd5bd1fe3095063d7fd8156d87e996f367f3/package.json#L50-L70

**Reply by m1sk9** [reply]
この鬱陶しい Warn を見なくて済む

![CleanShot 2024-07-20 at 03 52 42@2x](https://github.com/user-attachments/assets/eedb55f3-230e-424c-aeec-28086195037f)

**Comment by m1sk9** [comment]
@Allianaab2m , @laminne 

とりあえず Biome に置き換えた作業として https://github.com/pulsate-dev/pulsate/tree/replace-eslint-and-prettier-to-biome に Push しました. そのまますんなりマージできるよう VSCode 周りの設定も変えたのでローカルに落とせば実際に使えます. 

Biome の Formatter 及び Linter を走らせた状態で `./pkg` や `./script` は更新されていますが, その中で Biome がいくつか Lint でエラーを出しています. (前述の通り Lint のルールをそのまま Migrate はできなかったので recommend のルールをそのまま使っています)

いくつかのエラーは `biome check --write` で自動修正しましたが以下のエラーは修正できません.
これらのルールは受け入れるか無視するか考えたほうがいいと思ったのでリスト化しました. すり合わせをできればなと思います. 


### [noShadowRestrictedNames](https://biomejs.dev/linter/rules/no-shadow-restricted-names/)

Node.js のグローバル定数で使用されている変数名を Import 文で使用することを禁止します.

`mini-fn` の `Promise` に引っかかっているようです.

![CleanShot 2024-07-20 at 04 32 58](https://github.com/user-attachments/assets/915d0999-5d48-416b-8d63-85c80fdb68bd)

### [noNonNullAssertion](https://biomejs.dev/linter/rules/no-non-null-assertion/)

TypeScript の non-null assertion `!` を禁止します.

### [noUnreachableSuper](https://biomejs.dev/linter/rules/no-unreachable-super/) 

`super()` の複数使用を禁止する? (よくわからない)

**Reply by m1sk9** [reply]
@m1sk9 CI を置き換える（忘れてた）

**Reply by Allianaab2m** [reply]
検証ありがとうございます．
BiomeはLint/Formatエラーを無視する`biome-ignore`がありますが，これは必ず無視した理由を書かないとIgnoreされないということになっています．

cf. https://biomejs.dev/ja/linter/#%E3%82%B3%E3%83%BC%E3%83%89%E3%81%AE%E7%84%A1%E8%A6%96

よって，基本はルールを受け入れる方針で書き，必要に応じてIgnore + 理由の説明 + そのIgnoreが妥当かレビュー段階で検査の運用が好ましいと考えます．

`noShadowRestrictedNames`に関しては全部にIgnore書いてたんじゃ面倒なので外してもいいかなと．まあ，Biome的には別名importにしろってことなんでしょうけど．

**Reply by laminne** [reply]
`noNonNullAssertion`はちょっと微妙かも
もう直したかもしれないけど過去に型推論が諦めた部分で使っているので(TS5.5になったから使わないでよくなっている可能性はある)

**Reply by m1sk9** [reply]
`noUnreachableSuper` はちょっと読んでも理解できなかった. Biome 的に `super()` は一回のみ呼ばれることを想定しているのか?

**Reply by laminne** [reply]
`noUnreachableSuper`はESLintの`no-this-before-super`と同じらしい

**Reply by laminne** [reply]
`super()`を呼ぶなら`this`は`super()`よりもあとに書けというルールみたい

**Reply by m1sk9** [reply]
`noUnreachableSuper` はそのまま採用して `noShadowRestrictedNames` は無効かな... ( mini-fn を使ってる以上仕方ない)
`noNonNullAssertion` は現時点での型推論はやっぱりうまく行ってないからこれはコメントでの無効かな
![CleanShot 2024-07-20 at 15 11 49](https://github.com/user-attachments/assets/ca208200-75d7-4541-ad5a-0d55302943e2)

**Reply by m1sk9** [reply]
> `noUnreachableSuper` はそのまま採用して `noShadowRestrictedNames` は無効かな... ( mini-fn を使ってる以上仕方ない)
> 
> `noNonNullAssertion` は現時点での型推論はやっぱりうまく行ってないからこれはコメントでの無効かな
> 
> ![CleanShot 2024-07-20 at 15 11 49](https://github.com/user-attachments/assets/ca208200-75d7-4541-ad5a-0d55302943e2)
> 
> 

これで採用したいと思うけどどう？

**Reply by laminne** [reply]
とりあえずこれで良いと思います

**Comment by m1sk9** [comment]
PR を提出しました: #574
